### PR TITLE
added AutoCloseWorker for closing related workers when a page/another worker is closed.

### LINKED ISSUE #17977: Dedicated worker is not closed when associated document is destroyed
```html
<iframe src="http://localhost:8000/bar2.html" onload="this.remove()"></iframe>
```
```html
<script>
var w = new Worker('work.js');
</script>
```
```js
setInterval(function() {
  console.log('hi');
}, 500);
```
This shows an empty page (since the iframe is removed), but the worker continues logging to stdout.

### PR DESCRIPTION
**Pull request by nupurbaghel**
Followed steps as suggested in issue. 
Error is not yet resolved!

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #17977 (https://github.com/servo/servo/issues/17977?).

- [x] These changes do not require tests because it will have manual test case.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20068)

### CODE REVIEW

**Comment by nupurbaghel** [comment]
Wanted to confirm about the following -
1. Used Ordering::SeqCst for the drop function, also tried using Relaxed. What is more valid here?
2. Used the Default::default() for all places which were using GlobalScope::new_inherited. Is this fine?
3. Domrefcell< Vector > for global scope( emilio suggested RefCell < Option < > > ) Will it make any difference here ?

**Comment by jdm** [comment]
What if you add a method to GlobalScope that clears the vector, and call that from Window::clear_js_runtime?

File: components/script/dom/globalscope.rs [hunk 1]
```diff
@@ -124,6 +141,7 @@ impl GlobalScope {
         timer_event_chan: IpcSender<TimerEvent>,
         origin: MutableOrigin,
         microtask_queue: Rc<MicrotaskQueue>,
+        list_auto_close_worker: DomRefCell<Vec<AutoCloseWorker>>,
```
Comment (jdm) [inline components/script/dom/globalscope.rs]: We don't need to pass this in as an argument. We can simply initialize it with a default value instead.

File: components/script/dom/globalscope.rs [hunk 2]
```diff
@@ -142,9 +160,14 @@ impl GlobalScope {
             timers: OneshotTimers::new(timer_event_chan, scheduler_chan),
             origin,
             microtask_queue,
+            list_auto_close_worker,
         }
     }
 
+    pub fn update_list(&self, closing_worker: Arc<AtomicBool>) {
```
Comment (jdm) [inline components/script/dom/globalscope.rs]: Let's call this `track_worker` instead.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by asajeffrey** [comment]
For the ordering in the drop code, script is single-threaded, so you may be able to get away with just a `Cell<bool>`, and not need to worry about atomics.

**Comment by nupurbaghel** [comment]
@jdm When will the window call clear_js_runtime function ? Instead of passing the clone, we should do that ?
@asajeffrey I think we used atomic because we were adding a clone of closing from worker which was of atomic type. So should I try converting that now?

**Comment by jdm** [comment]
@nupurbaghel Window::clear_js_runtime is called from ScriptThread::handle_exit_pipeline_msg, which is called when the script thread receives a notification from the constellation that to close a pipeline. This happens when the constellation is notified that an iframe was removed, which is what happens in the testcase.

@asajeffrey The atomic boolean is shared with a worker thread, so that won't work.

**Comment by asajeffrey** [comment]
@jdm we have some uses of shared-memory concurrency in script? You learn something new every day.

**Comment by nupurbaghel** [comment]
I have added test file along with the suggested changes. I verified that the function to clear the vector gets called. The output of the test file still causes an endless printing of statements.  Can anything else be verified ?

**Comment by jdm** [comment]
Ah, the test file isn't quite right. I recommend the following:
parent.html
```html
<script>
var iframe = document.createElement('iframe');
iframe.src = "child.html";
iframe.onload = function() { iframe.remove(); }
document.body.appendChild(iframe);
</script>
```
child.html
```html
<script>
var w = new Worker('work.js');
</script>
```
work.js
```js
setInterval(function() {
  console.log('hi');
}, 500);
```
Does that change the behaviour for you?

**Comment by nupurbaghel** [comment]
Running the parent.html file raises the document.body is null error now :)

**Pull request merged**